### PR TITLE
feat(html pipe): add support for anchors on headings

### LINKED ISSUE #26: Add support for anchors on headings
Anchors are nicely managed on github.com: when rendering the markdown, github adds an anchor before each heading and shows an icon that gives the link to have a direct access to the corresponding heading.
While working on our Helix documentation, I realised this would be a really useful feature, especially to handle the navigation: you can have a long md file but create a nav that give direct access to a "section" of content below in the page. Without the anchors, I had to split the long md file into various small md files to give direct access.

The pipeline could automatically generate an anchor (`<a>` tag with an id) for each "section" and "heading". Showing or not an icon would be delegated to the project and could be done with CSS.

*This issue was moved by [trieloff](https://github.com/trieloff) from [adobe/project-helix#154](https://github.com/adobe/project-helix/issues/154).*

### PR DESCRIPTION
**Pull request by ramboz**
- Add a new `github-slugger` dependency to generate heading identifiers without collision
- Add unit some tests
- Introduce `heading-handler` in the `mdast-to-vdom`
- Adapt existing unit tests to support the heading identifier

Fixes #26

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -78,7 +78,8 @@
     "uri-js": "^4.2.2",
     "winston": "^3.0.0",
     "xml2js": "^0.4.19",
-    "xmlbuilder": "^12.0.0"
+    "xmlbuilder": "^12.0.0",
+    "xregexp": "^4.2.4"
```
Comment (tripodsan) [inline package.json]: is this really needed? can't you write the regexp w/o using this lib?

Comment (ramboz) [reply package.json]: Technically yes. This is what xregexp returns:
`/[^A-Za-zªµºÀ-ÖØ-öø-ˁˆ-ˑˠ-ˤˬˮͰ-ʹͶͷͺ-ͽͿΆΈ-ΊΌΎ-ΡΣ-ϵϷ-ҁҊ-ԯԱ-Ֆՙՠ-ֈא-תׯ-ײؠ-يٮٯٱ-ۓەۥۦۮۯۺ-ۼۿܐܒ-ܯݍ-ޥޱߊ-ߪߴߵߺࠀ-ࠕࠚࠤࠨࡀ-ࡘࡠ-ࡪࢠ-ࢴࢶ-ࢽऄ-हऽॐक़-ॡॱ-ঀঅ-ঌএঐও-নপ-রলশ-হঽৎড়ঢ়য়-ৡৰৱৼਅ-ਊਏਐਓ-ਨਪ-ਰਲਲ਼ਵਸ਼ਸਹਖ਼-ੜਫ਼ੲ-ੴઅ-ઍએ-ઑઓ-નપ-રલળવ-હઽૐૠૡૹଅ-ଌଏଐଓ-ନପ-ରଲଳଵ-ହଽଡ଼ଢ଼ୟ-ୡୱஃஅ-ஊஎ-ஐஒ-கஙசஜஞடணதந-பம-ஹௐఅ-ఌఎ-ఐఒ-నప-హఽౘ-ౚౠౡಀಅ-ಌಎ-ಐಒ-ನಪ-ಳವ-ಹಽೞೠೡೱೲഅ-ഌഎ-ഐഒ-ഺഽൎൔ-ൖൟ-ൡൺ-ൿඅ-ඖක-නඳ-රලව-ෆก-ะาำเ-ๆກຂຄງຈຊຍດ-ທນ-ຟມ-ຣລວສຫອ-ະາຳຽເ-ໄໆໜ-ໟༀཀ-ཇཉ-ཬྈ-ྌက-ဪဿၐ-ၕၚ-ၝၡၥၦၮ-ၰၵ-ႁႎႠ-ჅჇჍა-ჺჼ-ቈቊ-ቍቐ-ቖቘቚ-ቝበ-ኈኊ-ኍነ-ኰኲ-ኵኸ-ኾዀዂ-ዅወ-ዖዘ-ጐጒ-ጕጘ-ፚᎀ-ᎏᎠ-Ᏽᏸ-ᏽᐁ-ᙬᙯ-ᙿᚁ-ᚚᚠ-ᛪᛱ-ᛸᜀ-ᜌᜎ-ᜑᜠ-ᜱᝀ-ᝑᝠ-ᝬᝮ-ᝰក-ឳៗៜᠠ-ᡸᢀ-ᢄᢇ-ᢨᢪᢰ-ᣵᤀ-ᤞᥐ-ᥭᥰ-ᥴᦀ-ᦫᦰ-ᧉᨀ-ᨖᨠ-ᩔᪧᬅ-ᬳᭅ-ᭋᮃ-ᮠᮮᮯᮺ-ᯥᰀ-ᰣᱍ-ᱏᱚ-ᱽᲀ-ᲈᲐ-ᲺᲽ-Ჿᳩ-ᳬᳮ-ᳱᳵᳶᴀ-ᶿḀ-ἕἘ-Ἕἠ-ὅὈ-Ὅὐ-ὗὙὛὝὟ-ώᾀ-ᾴᾶ-ᾼιῂ-ῄῆ-ῌῐ-ΐῖ-Ίῠ-Ῥῲ-ῴῶ-ῼⁱⁿₐ-ₜℂℇℊ-ℓℕℙ-ℝℤΩℨK-ℭℯ-ℹℼ-ℿⅅ-ⅉⅎↃↄⰀ-Ⱞⰰ-ⱞⱠ-ⳤⳫ-ⳮⳲⳳⴀ-ⴥⴧⴭⴰ-ⵧⵯⶀ-ⶖⶠ-ⶦⶨ-ⶮⶰ-ⶶⶸ-ⶾⷀ-ⷆⷈ-ⷎⷐ-ⷖⷘ-ⷞⸯ々〆〱-〵〻〼ぁ-ゖゝ-ゟァ-ヺー-ヿㄅ-ㄯㄱ-ㆎㆠ-ㆺㇰ-ㇿ㐀-䶵一-鿯ꀀ-ꒌꓐ-ꓽꔀ-ꘌꘐ-ꘟꘪꘫꙀ-ꙮꙿ-ꚝꚠ-ꛥꜗ-ꜟꜢ-ꞈꞋ-ꞹꟷ-ꠁꠃ-ꠅꠇ-ꠊꠌ-ꠢꡀ-ꡳꢂ-ꢳꣲ-ꣷꣻꣽꣾꤊ-ꤥꤰ-ꥆꥠ-ꥼꦄ-ꦲꧏꧠ-ꧤꧦ-ꧯꧺ-ꧾꨀ-ꨨꩀ-ꩂꩄ-ꩋꩠ-ꩶꩺꩾ-ꪯꪱꪵꪶꪹ-ꪽꫀꫂꫛ-ꫝꫠ-ꫪꫲ-ꫴꬁ-ꬆꬉ-ꬎꬑ-ꬖꬠ-ꬦꬨ-ꬮꬰ-ꭚꭜ-ꭥꭰ-ꯢ가-힣ힰ-ퟆퟋ-ퟻ豈-舘並-龎ﬀ-ﬆﬓ-ﬗיִײַ-ﬨשׁ-זּטּ-לּמּנּסּףּפּצּ-ﮱﯓ-ﴽﵐ-ﶏﶒ-ﷇﷰ-ﷻﹰ-ﹴﹶ-ﻼＡ-Ｚａ-ｚｦ-ﾾￂ-ￇￊ-ￏￒ-ￗￚ-ￜ0-9²³¹¼-¾٠-٩۰-۹߀-߉०-९০-৯৴-৹੦-੯૦-૯୦-୯୲-୷௦-௲౦-౯౸-౾೦-೯൘-൞൦-൸෦-෯๐-๙໐-໙༠-༳၀-၉႐-႙፩-፼ᛮ-ᛰ០-៩៰-៹᠐-᠙᥆-᥏᧐-᧚᪀-᪉᪐-᪙᭐-᭙᮰-᮹᱀-᱉᱐-᱙⁰⁴-⁹₀-₉⅐-ↂↅ-↉①-⒛⓪-⓿❶-➓⳽〇〡-〩〸-〺㆒-㆕㈠-㈩㉈-㉏㉑-㉟㊀-㊉㊱-㊿꘠-꘩ꛦ-ꛯ꠰-꠵꣐-꣙꤀-꤉꧐-꧙꧰-꧹꩐-꩙꯰-꯹０-９ _-]+/g`

So I'd have to replace every one of these sequences by the proper unicode value… quite hard to maintain…

**Comment by tripodsan** [comment]
LGTM. but remove package-lock. thanks.

**Comment by ramboz** [comment]
@rofe I wish I could, but the linting rule is hardcoded to 2018. Should I update the linter as well as part of the PR?

**Comment by rofe** [comment]
I see... I think we should use the same template as in [helix-cli](https://github.com/adobe/helix-cli/blob/master/.eslintrc.js) everywhere...

**Comment by tripodsan** [comment]
> I see... I think we should use the same template as in [helix-cli](https://github.com/adobe/helix-cli/blob/master/.eslintrc.js) everywhere...

I just added it...

File: src/utils/heading-handler.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const fallback = require('mdast-util-to-hast/lib/handlers/heading');
+const HtmlId = require('./generate-html-id');
+
+let headingIdentifiersCache = {};
```
[the remark this answers is not in the document]
Comment (koraa) [reply src/utils/heading-handler.js]: Ok yeah, I added the test to testHTMLFromMarkdown (which would not be a bad place to test this anyway as it tests the entire pipeline)

File: src/utils/generate-html-id.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+const _ = require('lodash');
+
+module.exports = {
+
+  /**
+   * Generates a valid HTML identifier from the string
+   * @param {string} text The text to use to generate the identifier
+   * @returns {string} the string for the identifier
+   */
+  generateId: text => _.kebabCase(text),
```
[the remark this answers is not in the document]
Comment (ramboz) [reply src/utils/generate-html-id.js]: For me, there are 2 reasons to do this:

1. Abstract the `id` generation logic which helps refactoring. When I rewrote this 3 times in the PR, I only updated this single method, not the rest of the code
2. Expose the method directly so I can have dedicated unit tests to test edge cases without having to run the whole pipeline. I test default use cases with the pipeline, and edge cases directly on the method (tests are faster, and maybe some edge cases are not achievable through the normal flow)

File: src/utils/heading-handler.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramboz) [reply src/utils/heading-handler.js]: I didn't think about (1), so great point, and I agree I "assumed" (3) as I didn't find any examples against it in the current examples/tests I could find. But definitely great points.

I'll refactor accordingly

File: src/utils/heading-handler.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * @returns {string} The suffixed heading identifier
+ */
+function suffixHeadingIdentifier(headingIdentifier) {
+  // If heading is of the form `foo`, turn it to `foo-1`
+  // If heading is of the form `foo-1` already, turn it to `foo-1-1`
+  const headingIdentifierRoot = headingIdentifier.replace(/-\d+$/, '');
+  headingIdentifiersCache[headingIdentifier] = headingIdentifiersCache[headingIdentifier]
+    || (headingIdentifiersCache[headingIdentifierRoot] ? 1 : 0);
+
+  let suffix = '';
```
[the remark this answers is not in the document]
Comment (ramboz) [reply src/utils/heading-handler.js]: Yes. I didn't test these exact test cases, but similar ones. I'll just run this as well to make sure it doesn't emerge an edge case I did not consider 👍

File: src/utils/heading-handler.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  headingIdentifiersCache[headingIdentifier] = headingIdentifiersCache[headingIdentifier] || 0;
+  if (headingIdentifiersCache[headingIdentifier] > 0) {
+    suffix = `-${headingIdentifiersCache[headingIdentifier]}`;
+  }
+  headingIdentifiersCache[headingIdentifier] += 1;
+
+  return headingIdentifier + suffix;
+}
+
+function headingHandler() {
```
[the remark this answers is not in the document]
Comment (ramboz) [reply src/utils/heading-handler.js]: I'll take a look at your code and refactor accordingly!

**Comment by ramboz** [comment]
@koraa Thanks for the great feedback!

File: src/utils/heading-handler.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (koraa) [reply src/utils/heading-handler.js]: It really is a suggestion, not a requirment, but thanks for looking at my code. :)

File: src/utils/generate-html-id.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (koraa) [reply src/utils/generate-html-id.js]: Well, if you must leave it ;)

1. Normally I would wholeheartedly agree; it is a sign of a very good architecture to abstract things away so they can be reused! There is such a thing as overarchitecting though; each file and function and line of code we add adds a bit of overhead. Here we are adding an entire file for a single function that is called a single time and thename is a bit more nebulous than "kebabCase" and it's documentation doesn't really add much either; the intention of describing what the function is for is great btw, but the documentation doesn't really elaborate on what makes this function particularly great for that. I wholeheartedly agree that we should introduce abstraction, but maybe we start doing that when the function is used more than once or when it becomes more than an alias ;)
2. Yes and I think your tests are great. But so are lodash's and lodashs actually are a bit more elaborate and specific and they have been reviewed and used by a lot of people; they do contain tests specific to handling unicode (e.g. in test/words.js). I don't think we can approach that level of diligence or gain a lot by replicating that work – the chance of a bad regression in lodash that we can find with our tests is IMHO slim.

But I have spent far too much time elaborating on such a minor point and distracting from more important matters; if you decide to keep it that way, so be it! :)

File: src/utils/heading-handler.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tripodsan) [reply src/utils/heading-handler.js]: > One and two are likely to hold (>80% in the next couple of years) although (1) might break if we start doing partially async vdom conversions.

I think it is quite possible that we start using concurrent actions sometime. so better be prepared.

File: src/utils/generate-html-id.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ramboz) [reply src/utils/generate-html-id.js]: I'm actually good either way. I initially abstracted because the implementation was more complex and wanted the code to be re-usable (i.e. generate identifiers for other things than just headings), but as you pointed out, in the end with your improvements I'm just creating a proxy that doesn't add better tests nor documentation :)

**Comment by ramboz** [comment]
I'll continue working on this on Monday

**Comment by trieloff** [comment]
This package might be useful: https://github.com/remarkjs/remark-autolink-headings

**Comment by ramboz** [comment]
@trieloff Perfect! Thanks for the pointer. This will make it way easier

**Comment by trieloff** [comment]
That's the cool thing about remark/unified: for most problems there is already a module.

File: src/utils/mdast-to-vdom.js [hunk 6]
```diff
@@ -60,6 +61,9 @@ class VDOMTransformer {
       this._handlers[type] = (cb, node, parent) => VDOMTransformer.handle(cb, node, parent, that);
       return true;
     });
+
+    this._headingHandler = new HeadingHandler(options);
+    this.match('heading', this._headingHandler.handler());
```
Comment (ramboz) [inline src/utils/mdast-to-vdom.js]: As the heading handler needs to support resetting the "counters" for the headings, we need a slightly more complex construct for it

File: src/utils/mdast-to-vdom.js [hunk 7]
```diff
@@ -215,6 +219,14 @@ class VDOMTransformer {
     // create a JSDOM object with the hast surrounded by the provided tag
     return new JSDOM(`<${tag}>${VDOMTransformer.toHTML(this._root, this._handlers)}</${tag}>`).window.document.body.firstChild;
   }
+
+  /**
+   * Resets the transformer to avoid leakages between sequential transformations
+   */
+  reset() {
```
Comment (ramboz) [inline src/utils/mdast-to-vdom.js]: We need to be able to reset the transformer (i.e. the counters for the headings collision), so had to introduce a new method for it. Feel free to propose better naming if `reset` feels inappropriate

File: test/testHTMLFromMarkdown.js [hunk 8]
```diff
@@ -151,6 +151,25 @@ describe('Testing Markdown conversion', () => {
     );
   });
 
+  it('Quote with markdown', async () => {
```
Comment (ramboz) [inline test/testHTMLFromMarkdown.js]: Thanks @koraa For the added test case

File: test/testMdastToVDOM.js [hunk 9]
```diff
@@ -33,6 +33,9 @@ const assertTransformerYieldsDocument = (transformer, expected) => {
     new JSDOM(expected).window.document,
   );
 
+  // Reset the transformer between the 2 calls to avoid leakages in the tests
+  transformer.reset();
```
Comment (ramboz) [inline test/testMdastToVDOM.js]: Need to reset the transformer between transformations to avoid leakages in the tests

**Comment by ramboz** [comment]
@koraa I updated the PR according to your concerns:
- the heading handler is scoped to the transform instance now, so running parallel operations should not be an issue anymore
- all headings are now parsed, not just the direct children so the blockquote example works

**Comment by ramboz** [comment]
@trieloff Although I did use the github-slugger, I did not use the autolink one.
As mentioned in the issue itself (https://github.com/adobe/helix-pipeline/issues/26#issuecomment-480324004), I think this is an opinionated choice that should be project-specific.

File: package.json (hunk 1 above)
Comment (ramboz) [reply package.json]: Removed in favor of github-slugger module

**Comment by tripodsan** [comment]
@ramboz @trieloff : this contains a lot of intermediate commits.
who ever merges this. either rebase+squash to a single commit before performing a merge commit, or use the `squash and merge` button in github. please.

File: src/utils/heading-handler.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  /**
+   * Returns the handler function
+   */
+  handler() {
+    return (h, node) => {
+      // Prepare the heading id
+      const headingIdentifier = this.slugger.slug(HeadingHandler.getTextContent(node));
+
+      // Inject the id after transformation
+      const n = Object.assign({}, node);
```
Comment (koraa) [inline src/utils/heading-handler.js]: This totally is a side not and need not be addressed, but: `const n = {...node};`  :P

**Comment by ramboz** [comment]
@koraa I really appreciate the PR feedback, and since I'm still new to the project, I see all these comments as a great opportunity to both learn the architecture and best practices in the team.

I'm really not looking at having a 24h turn-around on my PRs at the moment, especially with me only working 20% on helix, and I'm happy if they take a couple of weeks at the moment.
